xds/src/main/java/io/grpc/xds/XdsLoadBalancer.java

```diff
@@ -145,30 +191,56 @@ public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
   private void handleNewConfig(XdsConfig xdsConfig) {
     String newBalancerName = xdsConfig.newBalancerName;
     LbConfig childPolicy = xdsConfig.childPolicy;
-    XdsComms xdsComms = null;
-    if (xdsLbState != null) { // may release and re-use/shutdown xdsComms from current xdsLbState
-      if (!newBalancerName.equals(xdsLbState.balancerName)) {
-        xdsComms = xdsLbState.shutdownAndReleaseXdsComms();
-        if (xdsComms != null) {
-          xdsComms.shutdownChannel();
-          xdsComms = null;
-        }
-      } else if (!Objects.equal(
-          getPolicyNameOrNull(childPolicy),
-          getPolicyNameOrNull(xdsLbState.childPolicy))) {
-        String cancelMessage = "Changing loadbalancing mode";
-        xdsComms = xdsLbState.shutdownAndReleaseXdsComms();
-        // close the stream but reuse the channel
-        if (xdsComms != null) {
-          xdsComms.shutdownLbRpc(cancelMessage);
-          xdsComms.refreshAdsStream();
-        }
-      } else { // effectively no change in policy, keep xdsLbState unchanged
-        return;
+    ManagedChannel lbChannel;
+    if (xdsLbState == null) {
+      lbChannel = initLbChannel(newBalancerName);
+      lrsClient =
+          lrsClientFactory.createLoadReportClient(lbChannel, helper, backoffPolicyProvider,
+              localityStore.getStatsStore());
+    } else if (!newBalancerName.equals(xdsLbState.balancerName)) {
+      if (lrsWorking) {
+        lrsClient.stopLoadReporting();
+        lrsWorking = false;
+      }
+      ManagedChannel oldChannel = xdsLbState.shutdownAndReleaseChannel("Changing balancer name");
+      oldChannel.shutdown();
+      lbChannel = initLbChannel(newBalancerName);
+      lrsClient =
+          lrsClientFactory.createLoadReportClient(lbChannel, helper, backoffPolicyProvider,
+              localityStore.getStatsStore());
+    } else if (!Objects.equal(
+        getPolicyNameOrNull(childPolicy),
+        getPolicyNameOrNull(xdsLbState.childPolicy))) {
+      lbChannel = xdsLbState.shutdownAndReleaseChannel("Changing load balancing mode");
+    } else { // effectively no change in policy, keep xdsLbState unchanged
+      return;
+    }
+    xdsLbState =
+        new XdsLbState(newBalancerName, childPolicy, helper, localityStore, lbChannel,
+            adsStreamCallback);
+  }
+
+  private ManagedChannel initLbChannel(String balancerName) {
```

Comment:
Better to take `Helper` in the argument and make this method static, so that it's clear it's a factory method and doesn't have any side effect.

Reply:
Done.